Cachepath work broken out into functional commits

rountree opened the pull request:
Fixes #61

rountree-alt committed:
Cachepath:  parse_loc utilities.

rountree-alt committed:
Cachepath:  remove/rename [orig_]location.

rountree-alt committed:
Cachepath:  Configure-time support.

rountree-alt committed:
Cachepath:  Internal messaging for path resolution

rountree-alt committed:
Cachepath:  Adds cobo_allreduce()

rountree-alt committed:
Cachepath:  Adds parameters to config_mgr

rountree-alt committed:
Cachepath:  adds flux parameter support

rountree-alt committed:
Cachepath:  Adds logging support.

rountree-alt committed:
Cachepath:  Removes out-of-root cleanup checks.

rountree-alt committed:
Cachepath:  Set of small, miscellaneous patches.

mplegendre commented on src/client/client/intercept_exec.c:
```diff
@@ -198,13 +196,12 @@ static char **updateEnvironment(char **envp, int *num_modified, int propogate_sp
       debug_printf2("Propogating spindle environment by copying it to new envp list\n");
       for (cur = (char **) envp; *cur; cur++, orig_size++);
-      new_size = orig_size + 10;
+      new_size = orig_size + 20;
```
Should be 'new_size = orig_size + 9'. Eight environment variables get propogated, plus one slot for the NULL.

rountree replied:
Fixed.

mplegendre commented on src/client/client/intercept_readlink.c:
```diff
@@ -31,19 +31,25 @@ Place, Suite 330, Boston, MA 02111-1307 USA
+    cachepath = chosen_realized_cachepath;
+    chosen_parsed_cachepath = chosen_parsed_cachepath;
+    chosen_symbolic_cachepath = chosen_symbolic_cachepath;
+}
```
I'm a bit confused by the set_intercept_readlink_cachepath() and set_should_intercept_cachepath(). Seems like we're keeping multiple copies of the paths in different static variables, then calling each of these functions to set the different copies.

Different variables holding the same copies of information are a source of bugs. Let's consolidate the variables holding paths in the clients to just one instance.

Or am I missing something where these are different?

rountree replied:
Two issues here:  
1) We're using setter functions to pass around pointers to static file-scoped variables.  Making the original variable global gets rid of the copies, but creates its own source of bugs.

2) We're keeping track of the realized, parsed, and symbolic versions of the cachepaths, Parsed and symbolic aren't used in `set_inercept_readlink_cachepath()` and symbolic isn't used in `set_should_intercept_cachepath()`.  (The code that references them in the setter function is there to silence the compiler warning.)

Let's err on the side of deleting more code.  Removed the two `set_*()` functions.

mplegendre commented on src/server/auditserver/ldcs_audit_server_process.h:
```diff
@@ -126,6 +126,11 @@ struct ldcs_process_data_struct
   ldcs_dist_model_t dist_model;
   ldcs_client_t* client_table;
   char *location;
```
We've still got 'location'. What's does 'location' represent anymore? Is it a duplicate variable for cachepath?

Also, please comment what these variables contain.

rountree replied:
The behavior of `location` hasn't changed, except that it's not longer used for cachepaths (it only affects the path used for fifo, the daemon, etc.).

We had discussed removing it entirely.  Let's leave that for a separate PR.

I've added comments.

mplegendre commented on src/include/spindle_launch.h:
```diff
@@ -126,6 +126,11 @@ typedef struct {
+   char *candidate_cachepaths;  /* Colon-separated list of candidate paths (max 64) */
+   char *chosen_cachepath;      /* The consensus path (same across all nodes). */
+   uint64_t cachepath_bitidx;   /* Bit index used by allReduce() to arrive at consensus. */
+
```
spindle_launch is the user interface function. We don't need to expose chosen_cachepath or cachepath_bitidx to the user. They're not selecting or reading those values.

Also, what's the difference between 'location' and 'candidate_cachepaths' here? What's it mean if a user sets both or only one?

rountree replied:
As above, `location` is used for paths to the fifo directory, the daemon file, and anything else that's not a cachepath.  I expect we'll remove it in a following PR.

Removing `chosen_cachepath` and `cachepath_bitidx` resulted in a modest string of perturbations, ending with relocating "parseloc.h" to `/src/util`.

mplegendre commented on src/client/client_comlib/client_api.c:
```diff
@@ -37,6 +37,59 @@ static struct lock_t comm_lock;
+   }
+
+   COMM_LOCK;
+   client_recv_msg_static(fd, &message, LDCS_READ_BLOCK);
```
Kind of messy doing one send and three recvs. In other parts of spindle the recv would be one message with three strings rather than three messages.

rountree replied:
Modified to use a single reply.  Removed future-proofing (not using the symbolic version of cachepaths).

mplegendre commented on src/client/client_comlib/client_api.c:
```diff
@@ -37,6 +37,59 @@ static struct lock_t comm_lock;
+   COMM_LOCK;
+   client_recv_msg_static(fd, &message, LDCS_READ_BLOCK);
+   COMM_UNLOCK;
+
```
If we keep this at one-send/three-recvs, then the locking here isn't correct. We're guarding against other threads in the same application process triggering spindle operations in parallel. This structure would allow another thread to grab the lock between the send/recvs and then recv the message meant for this thread.

Keep the lock through the entire send/recv stream (or condense this down to one recv message locked with the one send).

rountree replied:
Condensed to a single message.

mplegendre commented on src/client/client_comlib/client_api.c:
```diff
@@ -37,6 +37,59 @@ static struct lock_t comm_lock;
+
+   if (message.header.type != LDCS_MSG_CHOSEN_CACHEPATH || message.header.len > MAX_PATH_LEN) {
+      err_printf("Got unexpected message of type %d\n", (int) message.header.type);
+      assert(0);
```
As per our other commits, move away from assert(0) on network failure.

rountree replied:
Converted `assert(0)` to `return -1` in eight functions.

mplegendre commented on src/server/auditserver/ldcs_audit_server_handlers.c:
```diff
@@ -2936,6 +2943,79 @@ static int handle_client_pickone_msg(ldcs_process_data_t *procdata, int nc, ldcs
+        msgbundle_force_flush(procdata);
+    }
+
+    ldcs_audit_server_md_consensus(procdata, msg);
```
Suggest renaming ldcs_audit_server_md_consensus(...) to something like ldcs_audit_server_md_allreduce(MD_AND).

The ldcs_audit_server_md_* are all network operation focused, not spindle algorithm focused. That's the layer where you'd add a different network implementation (like infiniband), and we want to keep higher-level spindle concepts out of that layer.

rountree replied:
Renamed/redefined to `ldcs_audit_server_md_allreduce_AND( uint64_t *val )`.

I wanted to use your `MD_AND` suggestion, but that's going to result in a duplicated op enum in md-space and coho-space. As we're only using the bitwise-and op in md-space, let's kick that problem down the road and put the op name in the function name.  Ugly, but effective.

mplegendre commented on src/server/auditserver/ldcs_audit_server_handlers.c:
```diff
@@ -2936,6 +2943,79 @@ static int handle_client_pickone_msg(ldcs_process_data_t *procdata, int nc, ldcs
+   ldcs_send_msg(connid, &msg);
+   procdata->server_stat.clientmsg.cnt++;
+   procdata->server_stat.clientmsg.time += ldcs_get_time() - client->query_arrival_time;
+
```
The repeated adds to clientmsg.cnt and clientmsg.time is triple counting here.

rountree replied:
Resolved in earlier commit when we went to a single-message response.

mplegendre commented on src/flux/flux-spindle.c:
```diff
@@ -416,7 +416,8 @@ static int sp_getopts (flux_shell_t *shell, struct spindle_ctx *ctx)
                         "numa", &numa,
```
Similar comment to previous, what's the difference between location and cachepaths now?

rountree replied:
"location" will be renamed to "commpath" in a subsequent PR.

mplegendre commented on src/fe/startup/config_mgr.h:
```diff
@@ -125,7 +126,8 @@ enum CmdlineShortOptions {
    shortExecExcludes = 298,
-   shortPatchLdso
+   shortPatchLdso,
+   shortCachePaths,
```
Just to keep the pattern going, could you add numbers to the enum values (and fix shortPatchLdso while at it). I know it's not necessary, but it makes it easier to look values from debug_printfs if they're explicit here.

rountree replied:
Done.

mplegendre commented on src/cobo/cobo.c:
```diff
@@ -1432,6 +1432,48 @@ int cobo_allgather_str(char* sendstr, char*** recvstr, char** recvbuf)
+        /* not the root, so forward our reduction result to our parent */
+        if (cobo_write_fd(cobo_parent_fd, pval, sizeof(*pval)) < 0) {
+            err_printf("Sending reduced data to parent failed\n");
+            exit(1);
```
Don't exit on network failure

rountree replied:
Changed 27 instances of exiting on network failure to "return -1" and two to "return NULL".

Not confident that's a net gain.  If we end up pulling the network code out into its own library, we can revisit error handling then.

rountree-alt committed:
Fixes per Matt's comments.

rountree-alt committed:
Single source of truth for client cachepath.

Previously,

chosen_realized_cachepath was copied into
set_intercept_readlink_cachepath()

chosen_realized_cachepath and chosen_parsed_cachepath were copied
into set_should_intercept_cachepath()

This PR removes both setter functions and makes the original
pointers global.

rountree-alt committed:
Comments the cachepath variables.

rountree-alt committed:
Removes internal vars from spindle_launch.h

Removes chosen_cachepath and cachepath_bitindex from
  spindle_launch.h

Updates initialization of matching variables in ldcs_process_data.

determineValidCachePaths() moved from spindle_be.cc to
  ldcs_audit_server_process.c to get ldcs_process_data visibility.

Added #include "parseloc.h" to ldcs_audit_server_process.c to get
  declaration of determineValidCachePaths().

Relocated "parseloc.h" to src/util so ldcs_audit_server_process.c
  could find it.

Trued up signedness of types caused my making "parseloc.h" more
  visible, e.g., cachepath_bitidx is now uint64_t everywhere.

rountree-alt committed:
Client cachepath message now uses single response.

The three-message-reply response is now a single message with
two strings.  The symbolic version of the cachepath is no longer
communicated as it was not being used.

rountree-alt committed:
Removes assert(0) in network error paths.

rountree-alt committed:
Renames ldcs_audit_server_md_consensus().

New name is ldcs_audit_server_md_allreduce_AND().

If we get to the point where we're using other allreduce operations
we can solve the problem of duplicating the op list in md-land and
cobo-land.  For now, we're only using one op in md-land, so the
op can go into the function name.

rountree-alt committed:
Adds explicit enum values to CmdlineShortOptions.

rountree-alt committed:
Return instead of exit on network errors.

mplegendre commented on src/cobo/cobo.c:
```diff
@@ -160,7 +160,7 @@ static char* cobo_getenv(char* envvar, int type)
     char* str = getenv(envvar);
     if (str == NULL && type == ENV_REQUIRED) {
         err_printf("Missing required environment variable: %s\n", envvar);
-        exit(1);
```
Was this commit intended to be part of the cachepath work? It looks more to be part of the network reliability work.

And if there's a reason to do this in the cachepath PR, we would need to take a more robust pass to make sure the various error returns don't just lead to crashes.

mplegendre replied:
Oh, I see it's in response to my above comment. 

I'd more meant that we shouldn't propogate, not that clean them all up. We can move forward with this in the PR, but we'll want to come back and work through some of these new error return paths.

rountree replied:
Sounds good.

mplegendre commented on testsuite/test_driver.c:
```diff
@@ -1142,27 +1142,6 @@ static char* getCacheLocation(char *env_var)
    return strdup(last_slash);
 }
 
-static int checkLinkForLeak(const char *path, const char *spindle_loc)
```
This shouldn't be deleted. It's checking that spindle's readlink interception is working and that we don't have symlinks from /proc/PID/... pointing into the spindle cache location that can be seen by the application (which has happened before). 

We may need another way to get the cache path into the test, but we should keep the test.

rountree replied:
The cache path is available to the test, but the fifos are no longer in that path, so those files leak.  If I restore that code, these are the kind of errors I get on the first of the runTests:
```
 Error - [/p/vast1/rountree/repos/Spindle/testsuite/test_driver.c:1159] - Link at '/proc/self/fd/315' has path '/tmp/rountree/spindle.9d2b000000/spindle_comm/fifo-1379774-0', which leaks spindle path with 'spindle.9d2b000000'
 ```
I'll restore the `checkLinkForLeak` in the commpath PR and make sure that path gets into the `TestVerifier` class via another `<internal>` message (unless you'd rather I merge the commpath PR into this PR).  Just to be clear, the above file should NOT trigger an error, correct?

rountree replied:
Well, that was a good solution to a problem we don't have yet.
Here's the actual problem:

https://github.com/LLNL/Spindle/blob/d8ac4c58a0f2b4db00f4e420ccb7e608489bc687/testsuite/test_driver.c#L1241

Should be a bit simpler to solve.

mplegendre reviewed (commented):
Looks like the CI is failing on this PR due to unset TMPDIR changes. Resolve that and the deletion in test_driver.c and I think we'll be ready to merge.
